File: core/python/fedmcp/artifact.py
from datetime import datetime, timezone
from typing import Dict, Any, Optional
from uuid import UUID, uuid4
import json
import hashlib
from pydantic import BaseModel, Field, validator


class ArtifactType:
    """Standard artifact types from FedMCP spec v0.2"""
    SSP_FRAGMENT = "ssp_fragment"
    POAM_TEMPLATE = "poam_template" 
    AGENT_RECIPE = "agent_recipe"
    BASELINE_MODULE = "baseline_module"
    AUDIT_SCRIPT = "audit_script"
    
    # Extended types for healthcare
    RAG_QUERY = "rag_query"
    LLM_COMPLETION = "llm_completion"
    TOOL_INVOCATION = "tool_invocation"


class Artifact(BaseModel):
    """
    FedMCP Artifact as defined in spec v0.2
    
    All artifacts must be under 1 MiB when serialized
    """
    id: UUID = Field(default_factory=uuid4)
    type: str
    version: int = Field(default=1, ge=1)
    workspaceId: UUID = Field(alias="workspaceId")
    createdAt: str = Field(alias="createdAt")
    jsonBody: Dict[str, Any] = Field(alias="jsonBody")
    
    class Config:
        populate_by_name = True
    
    def __init__(self, **data):
        if "createdAt" not in data:
            data["createdAt"] = datetime.now(timezone.utc).isoformat() + "Z"
        super().__init__(**data)
    
    @validator("jsonBody")
    def validate_size(cls, v):
        """Ensure jsonBody doesn't exceed 1 MiB limit"""
        size = len(json.dumps(v).encode())
        if size > 1024 * 1024:
            raise ValueError(f"jsonBody size {size} exceeds 1 MiB limit")
        return v
    
    def canonicalize(self) -> bytes:
        """
        Return RFC 8785 canonical JSON representation
        """
        # For now, using standard JSON with sorted keys
        # TODO: Implement full RFC 8785 canonicalization
        return json.dumps(
            self.model_dump(by_alias=True),
            sort_keys=True,
            separators=(",", ":")
        ).encode()
    
    def hash(self) -> str:
        """Return SHA256 hash of canonical artifact"""
        return hashlib.sha256(self.canonicalize()).hexdigest()

File: core/typescript/src/artifact.ts
import { v4 as uuidv4 } from 'uuid';
import { createHash } from 'crypto';
import { Artifact } from './types';

export class FedMCPArtifact implements Artifact {
  id: string;
  type: string;
  version: number;
  workspaceId: string;
  createdAt: string;
  jsonBody: Record<string, any>;

  constructor(params: {
    type: string;
    workspaceId: string;
    jsonBody: Record<string, any>;
    version?: number;
    id?: string;
    createdAt?: string;
  }) {
    this.id = params.id || uuidv4();
    this.type = params.type;
    this.version = params.version || 1;
    this.workspaceId = params.workspaceId;
    this.createdAt = params.createdAt || new Date().toISOString();
    this.jsonBody = params.jsonBody;
    
    this.validate();
  }

  validate(): void {
    if (!this.id) throw new Error('Artifact ID is required');
    if (!this.type) throw new Error('Artifact type is required');
    if (!this.workspaceId) throw new Error('Workspace ID is required');
    if (this.version < 1) throw new Error('Version must be >= 1');
    if (!this.jsonBody || Object.keys(this.jsonBody).length === 0) {
      throw new Error('jsonBody cannot be empty');
    }
    
    // Check 1 MiB size limit
    const size = new TextEncoder().encode(JSON.stringify(this.jsonBody)).length;
    if (size > 1024 * 1024) {
      throw new Error(`jsonBody size ${size} exceeds 1 MiB limit`);
    }
  }

  canonicalize(): string {
    // RFC 8785 JSON Canonicalization
    // For now, using sorted keys
    return JSON.stringify(this, Object.keys(this).sort());
  }

  hash(): string {
    const canonical = this.canonicalize();
    return createHash('sha256').update(canonical).digest('hex');
  }

  toJSON(): Artifact {
    return {
      id: this.id,
      type: this.type,
      version: this.version,
      workspaceId: this.workspaceId,
      createdAt: this.createdAt,
      jsonBody: this.jsonBody
    };
  }
}